### PR TITLE
Moving most of mixer/doc/dev/development.md here

### PR DESCRIPTION
**Pull request by ldemailly**
https://github.com/istio/mixer/blob/master/doc/dev/development.md to
here with updates to make it a bit more generic than just mixer

### CODE REVIEW

File: devel/README.md [hunk 1]
```diff
@@ -1,7 +1,341 @@
...
+  - [Running the linters](#running-the-linters)
+  - [Running race detection tests](#running-race-detection-tests)
+  - [Adding dependencies](#adding-dependencies)
+  - [About testing](#about-testing)
+- [Local development scripts](#collection-of-scripts-and-notes-for-developing-for-istio)
+- [MacOS tips](#macos-tips)
+
+Other docs you should look at:
+
+- Each components additional development docs like [Mixer](https://github.com/istio/mixer/tree/master/doc/dev/development.md)'s
```
Comment (geeknoid) [inline devel/README.md]: How about providing an exhaustive list here?

File: devel/README.md [hunk 2]
```diff
@@ -1,7 +1,341 @@
...
+- [Go landmines](https://gist.github.com/lavalamp/4bd23295a9f32706a48f)
+- [Go style mistakes](https://github.com/golang/go/wiki/CodeReviewComments)
+
+## Prerequisites
+
+Istio components code bases has only few external dependencies you
```
Comment (geeknoid) [inline devel/README.md]: Istio components only have a few external dependencies

File: devel/README.md [hunk 3]
```diff
@@ -1,7 +1,341 @@
...
+need to setup before being able to build and run the code.
+
+### Setting up Go
+
+Mixer for instance is written in the [Go](http://golang.org) programming language.
```
Comment (geeknoid) [inline devel/README.md]: Let's generalize this.

"Many Istio components are written in..."

And let's just claim everything depends on Go 1.8. We'll do full org-wide upgrades when time comes.

File: devel/README.md [hunk 4]
```diff
@@ -1,7 +1,341 @@
...
+added to your ~/.profile:
+
+```shell
+export GOPATH=~/go
+export PATH=$PATH:$GOPATH/bin
+```
+
+### Setting up Bazel
+
+Istio components are built using the bazel build system. See
```
Comment (geeknoid) [inline devel/README.md]: bazel -> Bazel

File: devel/README.md [hunk 5]
```diff
@@ -1,7 +1,341 @@
...
+[here](https://bazel.build/versions/master/docs/install.html) for the
+installation procedures.
+
+### Setting up Docker
+
+To run some of Istio's examples and tests, you need to set up Docker server.
+Please follow [these instructions](https://docs.docker.com/engine/installation/)
+for how to do this for your platform.
+
+### Setting up personal access token
```
Comment (geeknoid) [inline devel/README.md]: Setting up a personal

File: devel/README.md [hunk 6]
```diff
@@ -1,7 +1,341 @@
...
+
+This is only necessary for core contributors / to push changes to the main repos. 
+You can make Pull Requests without it but the additional security is recommended for everyone.
```
Comment (geeknoid) [inline devel/README.md]: Pull Requests -> pull requests
without it -> without using an access token

File: devel/README.md [hunk 7]
```diff
@@ -1,7 +1,341 @@
...
+
+To be part of the Istio organization, we require two-factor authentication, and
+you must setup a personal access token to enable push via HTTPS. Please follow [these
+instructions](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/)
```
Comment (geeknoid) [inline devel/README.md]: I don't think markdown likes to split the link in two lines.

File: devel/README.md [hunk 8]
```diff
@@ -1,7 +1,341 @@
...
+bazel clean
+```
+### Running tests
+
+You can run all the available tests with:
+
+```shell
+bazel test ...
+```
+### Getting coverage numbers
```
Comment (geeknoid) [inline devel/README.md]: Does this apply to all repos? Same question of the other stuff like "make fmt" and friends. Those are setup for Mixer, and unless they've been copied elsewhere, other repos aren't on that plan (yet).

Comment (ldemailly) [inline devel/README.md]: it doesn't but I figured a developer will be smart to skip steps that don't work/apply as I mentioned in a few place we are using Mixer as an example component

it also doesn't apply to C++ for instance so we will need to refine those a bit in the future

File: devel/README.md (hunk 1 above)
Comment (ldemailly) [inline devel/README.md]: Well not every components has a specific dev guide and the list should be on
https://github.com/istio/istio/blob/master/README.md#repositories (though broker isn't...)

File: devel/README.md (hunk 2 above)
Comment (ldemailly) [inline devel/README.md]: ✅

File: devel/README.md (hunk 3 above)
Comment (ldemailly) [inline devel/README.md]: ✅

File: devel/README.md (hunk 4 above)
Comment (ldemailly) [inline devel/README.md]: ✅

File: devel/README.md (hunk 5 above)
Comment (ldemailly) [inline devel/README.md]: ✅  you reviewed the original ;-)

File: devel/README.md (hunk 6 above)
Comment (ldemailly) [inline devel/README.md]: ✅ with a small change (two-factor authentication instead of access token)

File: devel/README.md (hunk 7 above)
Comment (ldemailly) [inline devel/README.md]: you can't split the url but you can split the text of the link (it works, also not changed from earlier doc)

Comment (ldemailly) [inline devel/README.md]: it works

**Comment by ldemailly** [comment]
@geeknoid  thanks for the thorough review !
I think I covered all the notes, let me know if you want me to make further changes
(like listing components...)

File: devel/README.md [hunk 9]
```diff
@@ -1,7 +1,344 @@
...
+
+## Prerequisites
+
+Istio components only have few external dependencies you
+need to setup before being able to build and run the code.
+
+### Setting up Go
+
+Many Istio components are written in the [Go](http://golang.org) programming language.
+To build Mixer, you'll need a Go development environment. Builds for
```
Comment (geeknoid) [inline devel/README.md]: Mixer -> Istio in this paragraph.

**Pull request merged**